### PR TITLE
[FLINK-19636][coordination] Add DeclarativeSlotPool 

### PR DESCRIPTION
**Pull request by zentol**
Adds the DeclarativeSlotPool, a slot pool supporting declarative resource management.

This pool will for the time being not be used directly, as this will require proper integration into the scheduler. Instead, a wrapper will be added in FLINK-19314 that bridges the existing SlotRequest-based protocol to the declarative one.

### DeclarativeSlotPool

This pool no longer accepts slot requests but only in-/decrements to the requirements, which it accumulators, with the resulting total requirements being announced to the ResourceManager through a listener interface.

Slots that are offered by TaskExecutors are matched against these requirements; any slot not matching any requirement will be rejected.

The Scheduler then can make use of these slots by iterating over the free slots returned by the pool and reserving them.

The slot will remain reserved until it is either
a) freed by the Scheduler, in which case it will remain in the pool for future use
b) explicitly released, usually due to a failed allocated or TaskExecutor shutting down.

Freed slots remain in the pool so long as they correspond to a requirement; a slot that has been free for the idleTimeout and is required will be released. This action is periodically triggered from the outside.

As mentioned above, when slots are offered to the SlotPool they are matched against the declared requirements to figure out whether we actually need the slot.
However, this mapping between slots and the fulfilled requirement is mostly theoretical; the scheduler is free to use a slot to fulfill another requirement, e.g., because the locality of the slot makes it a desirable choice, despite potentially wasting resources.
To prevent deadlocks where the scheduler reserves slots in such a way that the requirements could no longer be fulfilled, the SlotPool automatically adjusts the requirements.
As an example, assume we have 2 types of resource profiles (small/large), where the large one can also be used for the small profile.
The job requires 1 of each, the slot pool is offered one of each, and it does a perfect match between the offered slots and requirements:
```
Requirements:
large: 1
small: 1

Mapping:
largeSlot <-> largeRequirement
smallSlot <-> smallRequirement
```

The scheduler may now decide to use the large slot to fulfill the small requirement. Without any further action, the job would not able to run since it is missing a large slot.
We prevent this by adjusting the requirements such that we need 1 more large slot (since we used up the other one) and one less small slot (since we fulfilled this requirement with the large one). The currently allocated small slot is hence no longer needed, and could be released:

```
Requirements:
large: 2
small: 0

Mapping:
largeSlot <-> smallRequirement
```

Once the second large slot is allocated and offered to the pool it will be matched against the outstanding large requirement, and the slot can start running.

### ResourceCounter

The ResourceCounter is an immutable version of the existing `org.apache.flink.runtime.slots.ResourceCounter`, it being immutable due to being exposed at various points in the API.

### AutoRequirementDecrementingSlotPoolWrapper

The current resource allocation protocol will be bridged to the declarative one by mapping SlotRequests to requirement increases, and slot freeing/releases to requirement reductions.
The increments will be done explicitly in the wrapper that we'll introduce in FLINK-19314, whereas the reductions will be handled in the `AutoRequirementDecrementingSlotPoolWrapper`.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPool.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	 *
+	 * <p>Whether the freed slot is returned to the owning TaskExecutor is implementation
+	 * dependent.
+	 *
+	 * @param allocationId allocationId identifying the slot to release
+	 * @param cause cause for releasing the slot; can be {@code null}
+	 * @param currentTime currentTime when the slot was released
+	 * @return info about freed slot, if a reserved slot with the given allocationId existed
+	 */
+	Optional<ResourceCounter> freeReservedSlot(AllocationID allocationId, @Nullable Throwable cause, long currentTime);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPool.java:116]: Optional is unnecessary; the returned ResourceCounter could just be empty

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPool.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	ResourceCounter releaseSlots(ResourceID owner, Exception cause);
+
+	/**
+	 * Releases the slot specified by allocationId if one exists.
+	 *
+	 * @param allocationId allocationId identifying the slot to fail
+	 * @param cause cause for failing the slot
+	 * @return info about released slot, if a slot with the given allocationId existed
+	 */
+	Optional<ResourceCounter> releaseSlot(AllocationID allocationId, Exception cause);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPool.java:134]: same as above

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+class AutoRequirementDecrementingSlotPoolWrapper implements DeclarativeSlotPool {
+
+	private final DeclarativeSlotPool backingSlotPool;
+
+	AutoRequirementDecrementingSlotPoolWrapper(DeclarativeSlotPool backingSlotPool) {
+		this.backingSlotPool = backingSlotPool;
+	}
+
+	@Override
+	public void increaseResourceRequirementsBy(ResourceCounter increment) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java:47]: It could make sense to instead pass a Collection<ResourceRequirement>, to stay in sync with the RM side of things

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+	private final Consumer<? super Collection<? extends PhysicalSlot>> notifyNewSlots;
+
+	private final Time idleSlotTimeout;
+	private final Time rpcTimeout;
+
+	private final AllocatedSlotPool slotPool;
+
+	private final Map<AllocationID, ResourceProfile> slotToRequirementProfileMappings;
+
+	private ResourceCounter totalResourceRequirements;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:83]: I would prefer it we would use the mutable ResourceCounter implementation here so the fields can be final, but as long as it's exposed in the API this would just create confusion.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+		for (AllocatedSlot allocatedSlot : allocatedSlots) {
+			allocatedSlot.releasePayload(cause);
+		}
+	}
+
+	@Override
+	public void releaseIdleSlots(long currentTimeMillis) {
+		final Collection<AllocatedSlotPool.FreeSlotInfo> freeSlotsInformation = slotPool.getFreeSlotsInformation();
+
+		//order?
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:351]:
```suggestion
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+	private final Consumer<? super Collection<ResourceRequirement>> notifyNewResourceRequirements;
+
+	private final Consumer<? super Collection<? extends PhysicalSlot>> notifyNewSlots;
+
+	private final Time idleSlotTimeout;
+	private final Time rpcTimeout;
+
+	private final AllocatedSlotPool slotPool;
+
+	private final Map<AllocationID, ResourceProfile> slotToRequirementProfileMappings;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:81]: It may make sense to replace this with a JobScopedResourceTracker to keep the matching logic in sync with the RM side.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+
+		Preconditions.checkState(allocatedSlot.getResourceProfile().isMatching(requiredSlotProfile), "");
+
+		ResourceProfile previouslyMatchedResourceProfile = Preconditions.checkNotNull(slotToRequirementProfileMappings.get(allocationId));
+
+		updateSlotToRequirementProfileMapping(allocationId, requiredSlotProfile);
+		// slots can be reserved for a requirement that is no in line with the mapping we computed when the slot was
+		// offered, so we have to adjust the requirements accordingly to ensure we still request enough slots to
+		// be able to fulfill the total requirements
+		adjustRequirements(previouslyMatchedResourceProfile, requiredSlotProfile);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:261]: this could be skipped if the profiles match

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+	@Override
+	public PhysicalSlot reserveFreeSlot(AllocationID allocationId, ResourceProfile requiredSlotProfile) {
+		final AllocatedSlot allocatedSlot = slotPool.reserveFreeSlot(allocationId);
+
+		Preconditions.checkState(allocatedSlot.getResourceProfile().isMatching(requiredSlotProfile), "");
+
+		ResourceProfile previouslyMatchedResourceProfile = Preconditions.checkNotNull(slotToRequirementProfileMappings.get(allocationId));
+
+		updateSlotToRequirementProfileMapping(allocationId, requiredSlotProfile);
+		// slots can be reserved for a requirement that is no in line with the mapping we computed when the slot was
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:258]:
```suggestion
		// slots can be reserved for a requirement that is not in line with the mapping we computed when the slot was
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+import javax.annotation.Nullable;
+
+import java.util.Collection;
+
+/**
+ * A wrapper around a {@link DeclarativeSlotPool} that automatically decrements requirements if a slot is freed or
+ * released.
+ */
+class AutoRequirementDecrementingSlotPoolWrapper implements DeclarativeSlotPool {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java:37]: It may actually be better to inline and decreases into the wrapper introduced in FLINK-19314; as it is we're just spreading things around, and create weird inconsistencies where the wrapper _does_ decrease requirements when releasing a pending requests, but does _not_ decrease requirements when releasing an allocated slot.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+		return previouslyFulfilledRequirement;
+	}
+
+	@Override
+	public ResourceCounter releaseSlots(ResourceID owner, Exception cause) {
+		ResourceCounter previouslyFulfilledRequirement = backingSlotPool.releaseSlots(owner, cause);
+
+		backingSlotPool.decreaseResourceRequirementsBy(previouslyFulfilledRequirement);
+		return previouslyFulfilledRequirement;
+	}
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java:95]: Why are we decreasing the requirements if `releaseSlots` is called? I think this should be called if a TM's heartbeat times out.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+	@Override
+	public Optional<ResourceCounter> releaseSlot(AllocationID allocationId, Exception cause) {
+		Optional<ResourceCounter> previouslyFulfilledRequirements = backingSlotPool.releaseSlot(allocationId, cause);
+		previouslyFulfilledRequirements
+			.ifPresent(backingSlotPool::decreaseResourceRequirementsBy);
+		return previouslyFulfilledRequirements;
+	}
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java]: Same here, shouldn't `releaseSlot` keep the requirements constant?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java (hunk 9 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java:37]: Yes, I think this might be cleaner.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @zentol. The changes look good to me. I had a couple of smaller comments. The one about the `slotToRequirementProfileMappings` is more of philosophical nature since I don't have a better idea right now. But we should keep this in mind and maybe we find a better solution. I agree with your comment that the introduced wrapper might actually be a bit more confusing than it helps.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+					taskManagerLocation,
+					taskManagerGateway);
+
+				acceptedSlots.add(allocatedSlot);
+				acceptedSlotOffers.add(matching.getSlotOffer());
+
+				acceptedResources = acceptedResources.add(matchedResourceProfile, 1);
+
+				// store the ResourceProfile against which the given slot has matched for future book-keeping
+				slotToRequirementProfileMappings.put(allocatedSlot.getAllocationId(), matchedResourceProfile);
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:174]: there are so many data-structures to modify (acceptedSlots/acceptedSlotOffers/acceptedResources), which would have to be passed in, along with other stuff (taskManagerLocation/Gateway), that it would become a bit of a mess.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+
+			if (unfulfilledResources.containsResource(slotOffer.getResourceProfile())) {
+				unfulfilledResources = unfulfilledResources.subtract(slotOffer.getResourceProfile(), 1);
+
+				// use the profile of the slot offer since it is a direct match against the requirement
+				matchingResourceProfile = slotOffer.getResourceProfile();
+			} else {
+				for (ResourceProfile unfulfilledResource : unfulfilledResources.getResources()) {
+					if (slotOffer.getResourceProfile().isMatching(unfulfilledResource)) {
+						matchingResourceProfile = unfulfilledResource;
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:204]: hmm...I think the idea was to accept any slot that could be used for _some_ requirement if they don't exactly match, to prevent cases where we reject a slot because some other slot has taken it's place. For example, assume the requirements are one small and one large slot. Large slot offer comes first, fulfills small requirement, small slot arrives and is rejected.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java (hunk 10 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java]: because we will request new slots for those that we just lost, and since this increments the requirements we have to decrement them somewhere to ensure the requirements stay the same.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+
+	@Override
+	public PhysicalSlot reserveFreeSlot(AllocationID allocationId, ResourceProfile requiredSlotProfile) {
+		final AllocatedSlot allocatedSlot = slotPool.reserveFreeSlot(allocationId);
+
+		Preconditions.checkState(allocatedSlot.getResourceProfile().isMatching(requiredSlotProfile), "");
+
+		ResourceProfile previouslyMatchedResourceProfile = Preconditions.checkNotNull(slotToRequirementProfileMappings.get(allocationId));
+
+		updateSlotToRequirementProfileMapping(allocationId, requiredSlotProfile);
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java]: I'm not sure if we can get _rid_ of it.

The information that _this_ slot was used to fulfill _that_ requirement is used for so many things; be it matching slot offers against outstanding requirements, figuring out whether a slot is eligible for an idleness timeout, or to update the fulfilled requirements if a slot is released.
It seems impossible to remove this mapping without removing everything that makes this pool more than a wrapper around the `AllocatedSlotPool`.

At the end of the day we need this information _somewhere_ anyway; be it in the slot pool, or the new scheduler so it could make sure to update the requirements.
Given that the declarative slot pool is all about managing slots in respects to requirements, it seems fitting that it contains such a mapping.

Maybe this is just a naming/structural issue? What if the DeclarativeSlotPool were not actually a pool, but a requirement manager?
Slot offers and releases would be issued directly against he AllocatedSlotPool, which informs the declarative slot pool, which then takes the appropriate steps of adjusting requirements etc. .
Basically, instead of removing the mapping portion out of the slot pool, remove the slot pool portion from it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:204]: I think the idea was not to accept excess slots which we are sure not to use. Which slots, if there are multiple matching ones, to accept is a different question. I think that can be a follow up question and I'd be fine to simply take the first matching one.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:174]: fair enough

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java (hunk 10 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java]: But shouldn't this be the responsibility of the user of this wrapper to do sort the requirements out correctly?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java]: I think my main concern comes from the following thought: I've considered the `DeclarativeSlotPool` as the component which makes sure that we keep a set of slots according to the declared resources and makes it accessible to a user for usage. If now the user decides to use a slot differently (e.g. for a smaller slot request), then it should be his responsibility to make sure that he no longer needs this slot or to adjust the resource requirements by asking for another big slot in order to fulfill the pending slot requests. Since this is something which has to be decided by the user, it shouldn't bother this component.

What I am not saying is that we cannot remember the mapping of a matching to speed up future matchings. This is perfectly fine and can be encapsulated in the implementation. So what bothers me is that the `reserveFreeSlot` method has a `requiredSlotProfile ` argument.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java (hunk 10 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java]: Currently the user of this wrapper does not know what mapping the slot pool came up with.
We would have to include the originally matched profile in the slot information, that is returned for free slots.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java]: or return it in `reserveFreeSlot`

Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java]: Ah I see. This is definitely a recurring theme that the matching of the slots to required resources and the slots to `Executions` need to be aligned.

In the case of the `releaseSlots` method, I guess that the return value will tell us which requirements are no longer fulfilled.

Maybe it is not such a bad idea to return in some form the matched resource profile of a slot. This would allow the user to adjust the resource requirements if it decides to use this slot for a different `Execution`.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AutoRequirementDecrementingSlotPoolWrapper.java]: hmm...it's not difficult to move the requirement adjustments into the wrapper, but the lack of updates to the slotpool-internal mappings could be problematic.

so let's see; we require one large and one small slot, get one each and do a perfect match.

```
requirements: S:1 L:1
fulfilledRequirements: S:1 L:1
slotToRequirementProfileMappings: S->S L->L
```

Now the scheduler makes things difficult, and uses L for S. We just adjust the requirements from the outside, and get:

```
requirements: S:2 L:0
fulfilledRequirements: S:1 L:1
mapping: S->S L->L
```

soo...we would now correctly determine that we need one more small slot, because we subtract the fulfilledRequirements from requirements, so that's fine.
But, if we release the large slot, then the wrapper needs to know what profile the slot was used for, to properly reduce the requirements. Currently this is based on what the slot pool returns, which would still based on the original mapping and as such would be incorrect.
I suppose if the scheduler were to use it then it would know for what requirement it was used for, so maybe the additional mapping would exist in the wrapper temporarily?...

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java (hunk 6 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:81]: Agreed, the resource tracker doesn't really fit; in particular because it cannot really handle the concept of slots being reserved.
I've moved the matching logic from it into a re-usable `RequirementMatcher` component that is now used by the tracker and slot pool.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+ * offered slots to the set of available and required resources and only accepts those
+ * slots which are required.
+ *
+ * <p>Slots which are released won't be returned directly to their owners. Instead,
+ * the slot pool implementation will only return them after the idleSlotTimeout has
+ * been exceeded by a free slot.
+ *
+ * <p>The slot pool will call {@link #notifyNewSlots} whenever newly offered slots are
+ * accepted or if an allocated slot should become free after it is being
+ * {@link #freeReservedSlot released}.
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:68]:
```suggestion
 * {@link #freeReservedSlot freed}.
```

bc release is used in the API and means that we remove the slot from the pool.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+	@Nonnull
+	private ResourceProfile getMatchingResourceProfile(AllocationID allocationId) {
+		return Preconditions.checkNotNull(slotToRequirementProfileMappings.get(allocationId), "No matching resource profile found for %s", allocationId);
+	}
+
+	@Override
+	public PhysicalSlot reserveFreeSlot(AllocationID allocationId, ResourceProfile requiredSlotProfile) {
+		final AllocatedSlot allocatedSlot = slotPool.reserveFreeSlot(allocationId);
+
+		Preconditions.checkState(allocatedSlot.getResourceProfile().isMatching(requiredSlotProfile), "");
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:226]: The error message is missing.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 17]
```diff
@@ -0,0 +1,428 @@
...
+		final Collection<AllocatedSlot> acceptedSlots = new ArrayList<>();
+
+		for (SlotOffer offer : offers) {
+			if (slotPool.containsSlot(offer.getAllocationId())) {
+				// we have already accepted this offer
+				acceptedSlotOffers.add(offer);
+			} else {
+				Optional<AllocatedSlot> acceptedSlot = matchOfferWithOutstandingRequirements(offer, taskManagerLocation, taskManagerGateway);
+				if (acceptedSlot.isPresent()) {
+					acceptedSlotOffers.add(offer);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:155]: We could log on debug which offers are matching with which requirements.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 18]
```diff
@@ -0,0 +1,428 @@
...
+		final AllocatedSlot allocatedSlot = slotPool.reserveFreeSlot(allocationId);
+
+		Preconditions.checkState(allocatedSlot.getResourceProfile().isMatching(requiredSlotProfile), "");
+
+		ResourceProfile previouslyMatchedResourceProfile = Preconditions.checkNotNull(slotToRequirementProfileMappings.get(allocationId));
+
+		if (!previouslyMatchedResourceProfile.equals(requiredSlotProfile)) {
+			// slots can be reserved for a requirement that is not in line with the mapping we computed when the slot was
+			// offered, so we have to update the mapping adjust the requirements accordingly to ensure we still request enough slots to
+			// be able to fulfill the total requirements
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:233]: Add debug logging which tells about the requirements adjustment.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 19]
```diff
@@ -0,0 +1,428 @@
...
+			updateSlotToRequirementProfileMapping(allocationId, requiredSlotProfile);
+			adjustRequirements(previouslyMatchedResourceProfile, requiredSlotProfile);
+		}
+
+		return allocatedSlot;
+	}
+
+	@Override
+	public ResourceCounter freeReservedSlot(AllocationID allocationId, @Nullable Throwable cause, long currentTime) {
+		LOG.debug("Release slot {}.", allocationId);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:243]: Let's be consistent with the terms and use `Free reserved slot {}.` here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 20]
```diff
@@ -0,0 +1,428 @@
...
+
+		final Optional<AllocatedSlot> releasedSlot = slotPool.freeReservedSlot(allocationId, currentTime);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:245]:
```suggestion
		final Optional<AllocatedSlot> freedSlot = slotPool.freeReservedSlot(allocationId, currentTime);
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java [hunk 21]
```diff
@@ -0,0 +1,428 @@
...
+		}
+
+		releaseSlots(slotsToReturnToOwner, new FlinkException("Returning idle slots to their owners."));
+	}
+
+	private void releaseSlots(Iterable<AllocatedSlot> slotsToReturnToOwner, Throwable cause) {
+		for (AllocatedSlot slotToReturn : slotsToReturnToOwner) {
+			Preconditions.checkState(!slotToReturn.isUsed(), "Free slot must not be used.");
+
+			LOG.info("Releasing slot [{}].", slotToReturn.getAllocationId());
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java:347]: maybe log on debug the `cause`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPoolTest.java [hunk 22]
```diff
@@ -0,0 +1,550 @@
...
+
+		slotPool.releaseSlots(taskManagerLocation.getResourceID(), new FlinkException("Test failure"));
+
+		final Collection<AllocationID> freedSlots = freeSlotConsumer.drainFreedSlots();
+
+		assertThat(freedSlots, containsInAnyOrder(slotOffers.stream().map(SlotOffer::getAllocationId).toArray()));
+	}
+
+	@Test
+	public void testFailSlotDecreasesResources() throws InterruptedException {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPoolTest.java:229]: Maybe better to call this test `testReleaseSlotDecreasesFulfilledResourceRequirements`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPoolTest.java [hunk 23]
```diff
@@ -0,0 +1,550 @@
...
+		final PhysicalSlot physicalSlot = physicalSlots.iterator().next();
+
+		slotPool.releaseSlot(physicalSlot.getAllocationId(), new FlinkException("Test failure"));
+
+		final ResourceCounter finalResourceRequirements = resourceRequirements.subtract(physicalSlot.getResourceProfile(), 1);
+		assertThat(slotPool.getFulfilledResourceRequirements(), is(finalResourceRequirements));
+	}
+
+	@Test
+	public void testFailSlotReturnsSlot() throws InterruptedException {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPoolTest.java:247]: `testReleaseSlotReturnsSlot`

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPoolTest.java [hunk 24]
```diff
@@ -0,0 +1,550 @@
...
+		slotPool.reserveFreeSlot(newSlot.getAllocationId(), RESOURCE_PROFILE_1);
+		slotPool.freeReservedSlot(newSlot.getAllocationId(), null, 0);
+
+		final Collection<PhysicalSlot> recycledSlots = drainNewSlotService(notifyNewSlots);
+
+		assertThat(Iterables.getOnlyElement(recycledSlots), sameInstance(newSlot));
+
+		final Collection<SlotOffer> newSlotOffers = createSlotOffersForResourceRequirements(initialRequirements);
+
+		// the pending requirement should be fulfilled by the released slot --> rejecting new slot offers
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPoolTest.java:357]:
```suggestion
		// the pending requirement should be fulfilled by the freed slot --> rejecting new slot offers
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPoolTest.java [hunk 25]
```diff
@@ -0,0 +1,550 @@
...
+		slotPool.decreaseResourceRequirementsBy(excessRequirements);
+
+		slotPool.releaseIdleSlots(offerTime + idleSlotTimeout.toMilliseconds());
+
+		assertThat(acceptedSlots, is(not(empty())));
+		assertThat(slotPool.getFulfilledResourceRequirements(), is(requiredResources));
+	}
+
+	@Test
+	public void testReleasedSlotWillBeUsedToFulfillOutstandingResourceRequirements() throws InterruptedException {
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPoolTest.java:337]:
```suggestion
	public void testFreedSlotWillBeUsedToFulfillOutstandingResourceRequirements() throws InterruptedException {
```

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @zentol. The changes look good to me. I only had a couple of minor comments. Please take a look.
